Source/WebCore/editing/FrameSelection.cpp

```diff
@@ -2146,11 +2146,17 @@ void FrameSelection::focusedOrActiveStateChanged()
     // Because Style::Resolver::checkOneSelector() and
     // RenderTheme::isFocused() check if the frame is active, we have to
     // update style and theme state that depended on those.
-    if (Element* element = m_document->focusedElement()) {
-        element->invalidateStyleForSubtree();
-        if (RenderObject* renderer = element->renderer())
+    if (RefPtr element = m_document->focusedElement()) {
+        do {
+            element->invalidateStyleForSubtree();
+            if (!element->isInShadowTree())
+                break;
+            element = element->containingShadowRoot()->host();
```

Comment:
Would it be better to call `containingShadowRoot` and null check it rather than checking `isInShadowTree` and relying on the relationship between these two checks?

Reply:
Also seems like a helper function for walking up to the containing shadow root's host could make this code a lot easier to read. That idiom must have come up elsewhere.

Reply:
Oh, actually we have such a function: `Node::shadowHost`. Will use that instead.